### ISSUE TITLE
There is no: field->type() == google::protobuf::FieldDescriptor::TYPE_MESSAGE

### ISSUE THREAD

**Issue by shadylabs** [zhanjunxiong/luapb#5]
I cant find anything that will deal with
" field->type() == google::protobuf::FieldDescriptor::TYPE_MESSAGE " 
in pb_set and pb_repeated_set functions. Is it intentionally or me just can't understand? I can't put fields of type another msg hence i found, that there is no code to deal with this type. I see such code in pb_get , pb_repeated_get  and pb_repeated_add thou.
Any help?

**Comment by zhanjunxiong** [comment]
you should use  xxx:add() for repeated message

**Comment by shadylabs** [comment]
I don't understand. Its not about the repeated message. Its about field of type MSG. Here is an example:

message RpbPutReq {
    required bytes bucket = 1;
    required RpbContent content = 2;
    optional bytes key = 3;
    optional bool return_body = 4;
}

message RpbContent {
    required bytes value = 1;
    optional bytes charset = 2;
    optional bytes content_encoding = 3;
}

The field content in first message is of type second message.
When i try to assign value /new RpbContent msg with valid value/  to content field of RpbPutReq i got error: bad argument #2 to '__newindex' (pb_set field_name type error)

**Comment by zhanjunxiong** [comment]
Can you give an example ？

**Comment by shadylabs** [comment]
yes, here it is:
wMSG = pb.new("RpbPutReq")  -- pb is local representation of LuaPB
cMSG = pb.new("RpbContent") -- the sub message
cMSG.value = "username" -- setting the required field "value" 
wMSG.bucket = "freecontent"  -- so far so good
wMSG.content= cMSG   ---- boom bad argument #2 to '__newindex' (pb_set field_name type error)
The error is obvious according to LuaPB.cc the check for the type in pb_set function has no if-then check for this type.

**Comment by shadylabs** [comment]
Is it right to assume that we should have equal number of types in all set/get/repeated_set/repeated_get/repeated_add functions?

**Comment by zhanjunxiong** [comment]
oh, i know

local wMsg = pb.new("RpbPutReq")
wMsg.bucket = "freecontent"
wMsg.content.value = "username"

or

you may be add a interface for copy_from or megre_from 

in my project, i use this.
example:
local wMsg = pb.new("RpbPutReq")
wMsg.bucket = "freecontent"
local cMSG = pb.new("RpbContent")
wMsg.content:copy_from(cMSG)

**Comment by shadylabs** [comment]
i thknk it just needs:
if(field->type() == google::protobuf::FieldDescriptor::TYPE_MESSAGE)
    {
    Message\* msg = reflection->MutableMessage(message, field);
    return push_message(L, msg, false);
    }

in the pb_set() and pb_repeated_set(). Isnt it?

**Comment by shadylabs** [comment]
Yeah i fixed the pb_set() function, now it seems there are more errors, pb_get() doesent work with msg type because of wrong method if i understand google's message.h header file

**Comment by zhanjunxiong** [comment]
i thknk it just needs:
if(field->type() == google::protobuf::FieldDescriptor::TYPE_MESSAGE)
{
Message\* msg = reflection->MutableMessage(message, field);
return push_message(L, msg, false);
}

in the pb_set() and pb_repeated_set(). Isnt it?

you can't do that

**Comment by zhanjunxiong** [comment]
I suggest you add a interface copy_from or merge_from

**Comment by shadylabs** [comment]
infact i just added in pb_set():
else if(field->type() == google::protobuf::FieldDescriptor::TYPE_MESSAGE)
        {
        Message\* msg = reflection->MutableMessage(message, field);
    return push_message(L, msg, false);
        }
and it works as a charm.

And with pb_repeated_set():

else if(field->type() == google::protobuf::FieldDescriptor::TYPE_MESSAGE)
        {
            Message\* msg = reflection->MutableRepeatedMessage(message, field,index);
        return push_message(L, msg,false);
        }

According to google headers there are GetMessage and GetRepeatedMessage  - getter functions, i believe intended to reside in pb_get() and pb_repeated_get() respectively but i have some type mismatch error so far.
You can't use MutableMessage in pb_get because it is actually a "Set" function.

**Comment by zhanjunxiong** [comment]
oh，maybe you are right！
You have a problem when you test？

**Comment by zhanjunxiong** [comment]
Thought for a moment, then do not realize is that in order to achieve simplicity.
If you implement an interface like you said, you want to deliver a “message” to c ++, and to manage what this “message” of gc. So I used  copy_from or megre_from alternatives such usage in my progject.
But you can try it yourself.
